### PR TITLE
ENH: CI benchmark runner

### PR DESCRIPTION
**Pull request by yarikoptic**
aims to address #3771 
- [x] splits into three dedicated actions (setup, run, compare)
- [ ] remove eval (see comment below)
- [ ] exits with non-0 in case of notable hit
- [ ] remove TEMP commit

### CODE REVIEW

**Comment by mih** [comment]
Note that some of the benchmarks are pretty fragile ('heavy_output' etc.), and have shown significant performance loss and gains (sometimes 35%). Those should not lead to benchmark failure, otherwise it becomes noise and will be ignored like appveyor.

**Comment by yarikoptic** [comment]
> Note that some of the benchmarks are pretty fragile ('heavy_output' etc.), and have shown significant performance loss and gains (sometimes 35%). Those should not lead to benchmark failure, otherwise it becomes noise and will be ignored like appveyor.

yeap, those should either be ignored (easier) or fixed up to become more robust (harder)

File: tools/ci/benchmark-travis-pr.sh [hunk 1]
```diff
@@ -23,20 +26,35 @@ run_asv () {
     pip install -e .
     git show --no-patch --format="%H (%s)"
     configure_asv
-    asv run -E existing --set-commit-hash $(git rev-parse HEAD)
+    asv run -E existing --set-commit-hash "$(git rev-parse HEAD)"
```
Comment (kyleam) [inline tools/ci/benchmark-travis-pr.sh]: Quotes aren't needed in this case.  (If you grep through git's source code you'll see that they usually do not surround `$(git rev-parse HEAD)` with quotes.)  But if it eases your mind or you like the consistency, it of course doesn't hurt.

File: tools/ci/benchmark-travis-pr.sh [hunk 2]
```diff
@@ -23,20 +26,35 @@ run_asv () {
...
-git checkout --force refs/bm/merge-target
-run_asv
+compare () {
+    asv compare refs/bm/merge-target refs/bm/pr
+}
 
-asv compare refs/bm/merge-target refs/bm/pr
+case "$action" in
+    '') setup && run && compare;;
+    *) eval "$action";;
```
Comment (kyleam) [inline tools/ci/benchmark-travis-pr.sh]: Please don't `eval` arbitrary input.  There are only a few valid cases, and I don't see any problem with having an arm for each one.

**Comment by yarikoptic** [comment]
hm, `compare` step still runs them, although they are presumably ran in a previous step ... not sure yet WTF

File: tools/ci/benchmark-travis-pr.sh (hunk 1 above)
Comment (yarikoptic) [reply tools/ci/benchmark-travis-pr.sh]: I succumb to pleasures of shellcheck. If there is no harm - I just obey

File: tools/ci/benchmark-travis-pr.sh (hunk 2 above)
Comment (yarikoptic) [reply tools/ci/benchmark-travis-pr.sh]: Yeah... Better be safe than sorry, will redo with a dedicated arm for all three and the eval ;-)

File: tools/ci/benchmark-travis-pr.sh (hunk 1 above)
Comment (kyleam) [reply tools/ci/benchmark-travis-pr.sh]: Fair enough.

**Comment by yarikoptic** [comment]
eh, still think this would be worthwhile, just need to get to it some time

**Comment by mih** [comment]
Closing due to inactivity.

**Pull request closed without merging**